price-feeder/oracle/provider/kraken.go

```diff
@@ -23,100 +26,385 @@ type (
+	var systemStatus KrakenEventSystemStatus
+	if err := json.Unmarshal(bz, &systemStatus); err != nil {
+		p.logger.Err(err).Msg("Kraken provider could not unmarshal KrakenEventSystemStatus")
+		return
+	}
+
+	if strings.EqualFold(systemStatus.Status, "online") {
+		return
+	}
+
+	p.keepReconnecting()
+}
+
+// setTickerPair sets an ticker to the map thread safe by the mutex.
+func (p *KrakenProvider) setTickerPair(symbol string, ticker TickerPrice) {
+	p.muTickers.Lock()
+	defer p.muTickers.Unlock()
+	p.tickers[symbol] = ticker
+}
+
+// ping to check websocket connection.
+func (p *KrakenProvider) ping() error {
+	return p.wsClient.WriteMessage(websocket.PingMessage, ping)
+}
+
+// subscribePairs write the subscription msg to the provider.
+func (p *KrakenProvider) subscribePairs(pairs ...string) error {
+	subsMsg := newKrakenSubscriptionMsg(pairs...)
+	return p.wsClient.WriteJSON(subsMsg)
+}
+
+// setSubscribedPairs sets N currency pairs to the map of subscribed pairs.
+func (p *KrakenProvider) setSubscribedPairs(cps ...types.CurrencyPair) {
+	p.muSubscribedPairs.Lock()
+	defer p.muSubscribedPairs.Unlock()
+
+	for _, cp := range cps {
+		p.subscribedPairs[cp.String()] = cp
+	}
+}
+
+// removeSubscribedTickers delete N pairs from the subscribed map.
+func (p *KrakenProvider) removeSubscribedTickers(tickerSymbols ...string) {
+	p.muSubscribedPairs.Lock()
+	defer p.muSubscribedPairs.Unlock()
+
+	for _, tickerSymbol := range tickerSymbols {
+		delete(p.subscribedPairs, tickerSymbol)
+	}
+}
+
+// toTickerPrice return a TickerPrice based on the KrakenTicker.
+func (ticker KrakenTicker) toTickerPrice(symbol string) (TickerPrice, error) {
+	if len(ticker.C) != 2 || len(ticker.V) != 2 {
+		return TickerPrice{}, fmt.Errorf("error converting KrakenTicker to TickerPrice")
+	}
+	// ticker.C has the Price in the first position
+	// ticker.V has the totla	Value over last 24 hours in the second position
+	return newTickerPrice("Kraken", symbol, ticker.C[0], ticker.V[1])
+}
+
+// newKrakenSubscriptionMsg returns a new subscription Msg.
+func newKrakenSubscriptionMsg(pairs ...string) KrakenSubscriptionMsg {
+	return KrakenSubscriptionMsg{
+		Event: "subscribe",
+		Pair:  pairs,
+		Subscription: KrakenSubscriptionChannel{
+			Name: "ticker",
+		},
+	}
+}
+
+// krakenPairToCurrencyPairSymbol receives a kraken pair formated
+// ex.: ATOM/USDT and return currencyPair Symbol ATOMUSDT.
+func krakenPairToCurrencyPairSymbol(krakenPair string) string {
+	return strings.Replace(krakenPair, "/", "", -1)
+}
+
+// currencyPairToKrakenPair receives a currency pair
+// and return kraken ticker symbol ATOM/USDT.
+func currencyPairToKrakenPair(cp types.CurrencyPair) string {
+	return strings.ToUpper(cp.Base + "/" + cp.Quote)
+}
+
+// normalizeKrakenPair checks if there is an anomaly
+// in the ticker, like the pair of BTC is returned as XBT.
+func normalizeKrakenPair(ticker string) string {
+	return strings.Replace(ticker, "XBT", "BTC", 1)
+}
```

Comment:
Not sure if this function is necessary - will we ever support BTC? 
If it is necessary we should be more specific in the description, since this does not check for any other anomalies : 
```suggestion
// normalizeKrakenBTCPair changes XBT pairs to BTC,
// since other providers list bitcoin as BTC 
func normalizeKrakenBTCPair(ticker string) string {
	return strings.Replace(ticker, "XBT", "BTC", 1)
}
```